express-zod-api/src/deep-checks.ts

```diff
@@ -96,56 +108,53 @@ export const hasNestedSchema = (
 
 export const hasUpload = (subject: IOSchema) =>
   hasNestedSchema(subject, {
-    condition: (schema) => schema._def[metaSymbol]?.brand === ezUploadBrand,
+    condition: (schema) =>
+      globalRegistry.get(schema)?.[metaSymbol]?.brand === ezUploadBrand,
     rules: {
       ...ioChecks,
-      [ezFormBrand]: (schema: FormSchema, { next }) =>
-        Object.values(schema.unwrap().shape).some(next),
+      [ezFormBrand]: ioChecks.object,
     },
   });
 
 export const hasRaw = (subject: IOSchema) =>
   hasNestedSchema(subject, {
-    condition: (schema) => schema._def[metaSymbol]?.brand === ezRawBrand,
+    condition: (schema) =>
+      globalRegistry.get(schema)?.[metaSymbol]?.brand === ezRawBrand,
     maxDepth: 3,
   });
 
 export const hasForm = (subject: IOSchema) =>
   hasNestedSchema(subject, {
-    condition: (schema) => schema._def[metaSymbol]?.brand === ezFormBrand,
+    condition: (schema) =>
+      globalRegistry.get(schema)?.[metaSymbol]?.brand === ezFormBrand,
     maxDepth: 3,
   });
 
 /** @throws AssertionError with incompatible schema constructor */
-export const assertJsonCompatible = (subject: IOSchema, dir: "in" | "out") => {
-  const lazies = new WeakSet<z.ZodLazy<z.ZodTypeAny>>();
+export const assertJsonCompatible = (subject: $ZodType, dir: "in" | "out") => {
+  const lazies = new WeakSet<() => $ZodType>();
```

Comment:
interfaces can also be circular. should probably handle it similar to zts implementation